File: src/jade/semantic/cube_client.py
"""Cube.js semantic layer client.

Fail-fast: connection errors propagate immediately.
Uses fake client for testing.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class CubeConfig:
    """Configuration for the Cube.js client."""

    api_url: str
    api_key: str

    def __post_init__(self) -> None:
        if not self.api_url or not self.api_url.strip():
            msg = "api_url must be a non-empty string"
            raise ValueError(msg)
        if not self.api_key or not self.api_key.strip():
            msg = "api_key must be a non-empty string"
            raise ValueError(msg)


class CubeClient:
    """Cube.js semantic layer query client."""

    def __init__(self, config: CubeConfig, *, use_fake: bool = False) -> None:
        self._config = config
        self._use_fake = use_fake
        self.api_url = config.api_url

    def query(self, query: dict[str, Any]) -> dict[str, Any]:
        """Execute a Cube.js JSON Query."""
        if not query.get("measures") and not query.get("dimensions"):
            msg = "Query must have at least 'measures' or 'dimensions'"
            raise ValueError(msg)

        if self._use_fake:
            return self._fake_query(query)

        msg = "Real Cube.js API not implemented. Use use_fake=True for testing."
        raise NotImplementedError(msg)

    def _fake_query(self, query: dict[str, Any]) -> dict[str, Any]:
        """Return fake query results for testing."""
        measures = query.get("measures", [])
        dimensions = query.get("dimensions", [])

        row: dict[str, Any] = {}
        for m in measures:
            row[m] = 42
        for d in dimensions:
            row[d] = "sample_value"

        return {"data": [row]}

File: ts/semantic/cube-client.ts
/**
 * Cube.js semantic layer client.
 *
 * Fail-fast: connection errors throw immediately.
 */

export interface CubeConfig {
  readonly apiUrl: string;
  readonly apiKey: string;
  readonly useFake?: boolean;
}

export class CubeClient {
  public readonly apiUrl: string;
  private readonly useFake: boolean;

  constructor(config: CubeConfig) {
    if (!config.apiUrl || !config.apiUrl.trim()) {
      throw new Error("apiUrl must be a non-empty string");
    }
    if (!config.apiKey || !config.apiKey.trim()) {
      throw new Error("apiKey must be a non-empty string");
    }
    this.apiUrl = config.apiUrl;
    this.useFake = config.useFake ?? false;
  }

  query(query: Record<string, any>): { data: Record<string, any>[] } {
    if (!query.measures && !query.dimensions) {
      throw new Error("Query must have at least 'measures' or 'dimensions'");
    }

    if (this.useFake) {
      return this.fakeQuery(query);
    }

    throw new Error(
      "Real Cube.js API not implemented. Use useFake: true for testing."
    );
  }

  private fakeQuery(
    query: Record<string, any>
  ): { data: Record<string, any>[] } {
    const measures: string[] = query.measures ?? [];
    const dimensions: string[] = query.dimensions ?? [];

    const row: Record<string, any> = {};
    for (const m of measures) {
      row[m] = 42;
    }
    for (const d of dimensions) {
      row[d] = "sample_value";
    }

    return { data: [row] };
  }
}
